rust/kernel/platform.rs

```diff
@@ -180,6 +181,36 @@ impl Device {
         // SAFETY: By the type invariants, we know that `self.ptr` is non-null and valid.
         unsafe { (*self.ptr).id }
     }
+
+    /// Gets a system resources of a platform device.
+    pub fn get_resource(&self, rtype: IoResource, num: usize) -> Option<Resource> {
```

Comment:
One thing we do in `amba` and probably should do here too is to "remember" whether a resource was already taken. The idea is to limit the usage of a resource to only once so that we can avoid, for example, violating Rust's aliasing rules. See `amba::Device::take_resource`.

Can we do something similar here?

Reply:
> One thing we do in `amba` and probably should do here too is to "remember" whether a resource was already taken. The idea is to limit the usage of a resource to only once so that we can avoid, for example, violating Rust's aliasing rules. See `amba::Device::take_resource`.
> 

I see, this seems like a good idea. Some of the resource need to be shared across drivers - 
that is something that needs to be consider, but here this covers Rust's aliasing rules of the single module
if I understand correctly?

> Can we do something similar here?

I'll give a try to implement it.

Reply:
> I noticed that in `bindings::amba_device`, the device data structure already contains `resource` data. However, in `bindings::platform_device`, the device data structure contains a pointer to `resource` data. So, the implementation of platform driver should give `index` when calling `ioremap_resource()`. This means the abstraction may differ from `amba`.
> 

Yeah, but essentially the issue remains the same that is to ensure proper aliasing of retrieved resources.

> Moreover, I noticed that the `io_mem::IoMem` data structure must be initialized with a const `SIZE`. I think it only fits the model of `amba` rather than `platform`, as we need to choose resources by `index` in the runtime.
> 

I think that `SIZE` has nothing to do with this issue.

> In a word, I think it might need more changes to implement this.

Yeah, it's harder the fact it's necessary to check if the given `(RES_TYPE, index)` pair was used previously.

Reply:
> I think that SIZE has nothing to do with this issue.

The driver may be compatible to multiple devices, and each of them may have different size of resource. So you don't know the size of the resource you get in the compile time.

Reply:
> The driver may be compatible to multiple devices, and each of them may have different size of resource. So you don't know the size of the resource you get in the compile time.

`SIZE` is the minimum size. This is not an issue.

> Yeah, it's harder the fact it's necessary to check if the given `(RES_TYPE, index)` pair was used previously.

What I was thinking for this was to have a 64-bit number for each resource type. Each bit indicates whether the resource was previously taken. This obviously limits the number of resources to 64 per type but it's a simple implementation and covers the vast majority of devices.

Reply:
> Sure, I was thinking about taking advantage that each resource has its unique position in the array so that its array index may be calculated using pointer difference and then masked. WDYT?

I like this idea. I think it's simpler than mine because it requires a single mask. (Perhaps more fragile because it would break if the underlying implementation changed, but we can detect that the pointer is within the range we expect.)

Anyway, I'm happy with your proposed solution.

Reply:
@wedsonaf I provided the solution with with a single 64-bit mask. That should be fine for smaller drivers. In future this could be replaced with wider `bitmap<SIZE>` or similar. May you please review?